examples/bipXYZ.c

```diff
@@ -0,0 +1,390 @@
+#include <stdio.h>
+#include <assert.h>
+#include <string.h>
+
+#include "examples_util.h"
+
+#include <secp256k1.h>
+#include <secp256k1_extrakeys.h>
+#include <secp256k1_schnorrsig.h>
+
+
+static const unsigned char bipXYZ_tag[] = {'B', 'I', 'P', '0', 'X', 'Y', 'Z', '/', 'n', 'o', 'n', 'c', 'e'};
+static const unsigned char ZERO_MASK[32] = {
+      84, 241, 105, 207, 201, 226, 229, 114,
+     116, 128,  68,  31, 144, 186,  37, 196,
+     136, 244,  97, 199,  11,  94, 165, 220,
+     170, 247, 175, 105, 39,  10, 165,  20
+};
+
+
+static int schnorrsig_noncefp_bipXYZ(const secp256k1_context* ctx, unsigned char *nonce32, unsigned char *Q_ser, const unsigned char *msg32, unsigned char *seckey, unsigned char *nonce_commit) {
+    secp256k1_keypair keypair;
+    secp256k1_keypair kp;
+    secp256k1_xonly_pubkey Q;
+    secp256k1_xonly_pubkey X;
+    unsigned char masked_key[32];
+    unsigned char tweak[32];
+
+    unsigned char tweak_ser[96];
+    unsigned char q_ser[96];
+    int rv, pk_parity, i;
+
+    rv = secp256k1_keypair_create(ctx, &keypair, seckey);
+    assert(rv);
+    rv = secp256k1_keypair_xonly_pub(ctx, &X, &pk_parity, &keypair);
+    assert(rv);
+    if (pk_parity == 1) {
+        rv = secp256k1_ec_seckey_negate(ctx, seckey);
+        assert(rv);
+    }
+
+    /* t = masked secret key x */
+    for (i = 0; i < 32; i++) {
+        masked_key[i] = seckey[i] ^ ZERO_MASK[i];
```

Comment:
could we probably use` nonce_commit` itself to xor secret key ? in BIP0340 they use `aux_rand` which is similarly provided by user. 

meta question: do we need to mask it at all as it is going to hash function ?

Reply:
I'm guessing the "masking" is equivalent to a tagged hash just faster. I would use either one of them.

Reply:
in [BIP0340](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki#default-signing) they do both mask + tagged hash

Reply:
From what I see, the masked_key contains the bip340_sha256_tagged(data) ^ key_32, because if there is no data, the tagged hash can be precomputed.

Normally we define the deterministic nonce (simplified) like this:
k = H(x,m,n)

BIP-340 does something to this effect in two steps:
nx = H<sub>aux</sub>(n) ^ x
k = H<sub>nonce</sub>(nx|X|m)

The reason for this is an optimization, and also because m can be arbitrarily long. In both tagged hashes the midstate is also precomputed for performance. Pretty sleek.

My first guess was wrong.

Reply:
I see no reason why we couldn't generate `q` exactly like BIP-340 nonce gen.
The thing is the branch that has the ZERO_MASK should never come to play!!
We will always have `n` nonce extra they call `data`.

Reply:
make sense https://github.com/scgbckbone/secp256k1/pull/1/commits/cd888ef129bdf59c07aaaa0ee2b6c49bd830e37f